### PR TITLE
Throw error if user attempts to change an unchangeable setting

### LINKED ISSUE #1027: Setting a non-runtime config item should produce a warning
**Describe the bug**
In the late 90's I did some toying around with various demoscene effect. One of these effects was a copper bar. I made several small executables that displayed one or more horizontal color bars moving in a smooth up/down motion.

I coded these on (I think) a 486 with a Trident VGA card, and I'm certain they worked on that machine on pure DOS.

However, they do not display anything at all on DosBox-staging.

**To Reproduce**
Run the attached executables in DosBox

**Expected behaviour**
some fancy colored bars should appear, and maybe move up or down the screen (depending on the exe file... I don't remember what effect each EXE file does, it's been over 20 years by now...)

**Environment (please complete the following information):**
- Windows 7
- x86_64

**Version of dosbox-staging:**
- dosbox-staging version 0.77.0-alpha-760-gdc029

**Configuration parameters:**
Basically I tried svga_se, vgaonly, svga_paradise and some others. I also tried the `vertical retrace poll debug line` parameter but it didn't show anything

**Additional context**
I'm not sure my code is (technically) correct... but it did work on real hardware so I'm not sure if this qualifies as a bug or not...

The programs are pretty small and straightforward. I will try and see if I can find the source code, but basically it's more or less waiting for bit 3 of 0x3da, writing to 0x3c8, and waiting for bit 1 of 0x3da.
[copper.zip](https://github.com/dosbox-staging/dosbox-staging/files/6528684/copper.zip)

### LINKED ISSUE #3080: Better handling for config settings that can/cannot be changed at runtime
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

latest main

### What Operating System are you using?

Linux x86_64

### If Other OS, please describe

Arch Linux

### Is your feature request related to a problem? Please describe.

I was looking at issue #2527 where setting `fullscreen=true` does not work at runtime.  Looking at the code, we have an enum that describes when a particular config setting is allowed to be changed.

```cpp
struct Changeable {
	enum Value { Always, WhenIdle, OnlyAtStart, Deprecated };
};

Pbool = sdl_sec->Add_bool("fullscreen", always, false);

pbool = sdl_sec->Add_bool("pause_when_inactive", on_start, false);
```

Here we can see the usage code of two values.  `fullscreen` is meant to always be changeable while `pause_when_inactive` is meant to not be changeable at runtime.

However, I can't find real enforcement of this nor any documentation to the user.  Ideally, typing in `pause_when_inactive=true` would present the user with an error indicating that this value cannot be changed.  However, as it currently stands, there is no error and the window actually flickers because it is running some re-init code.  We shouldn't be trying to refresh the window when the user wants to change an unchangeable value.

Furthermore, it actually seems to update the config setting to the new value even though the pause functionality is unchanged.

pause

### Describe the solution you'd like

Unchangeable values should report an error when the user tries to change them.  We should not be running init functions when this happens.  We should not be updating the value at all.

It would also be nice if this was documented somewhere.  Perhaps in the config file, we write out when this setting is allowed to be changed.

### Describe alternatives you've considered

I don't know but I am open to suggestions.

### Add any other context or screenshots about the feature request here.

_No response_

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

### PR DESCRIPTION
**Pull request by weirddan455**
# Description

The changeable when value on config settings was not actually being enforced anywhere.  This is a simple change to the `config -set` DOS program to check that value and throw an error if the setting is not meant to be changed at runtime.

## Related issues

Fixes #3080
Fixes #1027

# Manual testing

`pause_when_inactive=true` no longer changes the value or runs destroy + init functions.  It instead simply throws an error stating the setting is not allowed to be changed.

Tested with several other config settings as well.

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [ ] commented on the particularly hard-to-understand areas of my code.
- [ ] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

**Comment by rderooy** [comment]
Just thinking, perhaps if such a case DOSBox Staging should give the option to "reboot" the emulator with the changed setting.

**Comment by weirddan455** [comment]
> Just thinking, perhaps if such a case DOSBox Staging should give the option to "reboot" the emulator with the changed setting.

Yeah there's another open issue #2565 requesting that feature.  I suppose it wouldn't be too hard to implement.  The restart feature literally restarts the entire process so it would just be a matter of launching the new process like `dosbox --set propery=value`.

File: src/misc/setup.cpp [hunk 1]
```diff
@@ -867,6 +867,16 @@ Property* Section_prop::Get_prop(int index)
...
 }
 
+Property* Section_prop::Get_prop(const std::string& propname)
+{
+	for (Property* property : properties) {
+		if (property->propname == propname) {
+			return property;
+		}
+	}
+	return nullptr;
```
Comment (japsmits) [inline src/misc/setup.cpp:877]:
```suggestion
    const auto it = std::ranges::find_if(properties, [propname](const auto* prop) { return prop->propname == propname); };
    return it != properties.end() ? *it : nullptr;
```

File: src/misc/programs.cpp [hunk 2]
```diff
@@ -798,7 +798,20 @@ void CONFIG::Run(void)
 			if (strlen(result)) {
 				WriteOut(result);
 			} else {
-				Section* tsec = control->GetSection(pvars[0]);
+				Section_prop* tsec = dynamic_cast<Section_prop *>(control->GetSection(pvars[0]));
```
Comment (japsmits) [inline src/misc/programs.cpp:801]:
```suggestion
				auto* tsec = dynamic_cast<Section_prop *>(control->GetSection(pvars[0]));
```

File: src/misc/programs.cpp [hunk 3]
```diff
@@ -798,7 +798,20 @@ void CONFIG::Run(void)
...
+				if (!tsec) {
+					WriteOut(MSG_Get("PROGRAM_CONFIG_PROPERTY_ERROR"), pvars[0].c_str());
+					return;
+				}
+				Property* property = tsec->Get_prop(pvars[1]);
```
Comment (japsmits) [inline src/misc/programs.cpp:806]:
```suggestion
				const auto* property = tsec->Get_prop(pvars[1]);
```

File: src/misc/setup.cpp (hunk 1 above)
Comment (weirddan455) [reply src/misc/setup.cpp:877]: I think the loop is much easier to read.

**Review by johnnovak — APPROVED** [review]
I like this change, nice job @weirddan455.

**Comment by weirddan455** [comment]
I was just thinking through the details of how setting these on restart would work.  There's a couple ways we could go about it:

1. After the "cannot be changed at runtime" error, prompt the user if they would like to restart.  This may break some people's .bat files that aren't expecting an interactive prompt and also would be only changing one setting at a time.
2. As @johnnovak said in #2565 we could maintain a list of settings to change on next restart, either by `config - r` or by the `ctrl + alt + home` hotkey.  I don't think we would need the duplicate the entire config data structure, just a string with key + value pairs should suffice to append onto the command line arguments of the new process.
3. Just merge this PR as is.

Btw, you already kind of can use `config -r` to achieve this.  That command lets you specify command line arguments.  You can do, for example, `config -r --set machine=tandy`.

**Comment by johnnovak** [comment]
> I was just thinking through the details of how setting these on restart would work. There's a couple ways we could go about it:
> 
>     1. After the "cannot be changed at runtime" error, prompt the user if they would like to restart.  This may break some people's .bat files that aren't expecting an interactive prompt and also would be only changing one setting at a time.
> 
>     2. As @johnnovak said in [Changing machine type or mem settings with config command will take effect on the next restart with config -r #2565](https://github.com/dosbox-staging/dosbox-staging/issues/2565) we could maintain a list of settings to change on next restart, either by `config - r` or by the `ctrl + alt + home` hotkey.  I don't think we would need the duplicate the entire config data structure, just a string with key + value pairs should suffice to append onto the command line arguments of the new process.
> 
>     3. Just merge this PR as is.
> 
> 
> Btw, you already kind of can use `config -r` to achieve this. That command lets you specify command line arguments. You can do, for example, `config -r --set machine=tandy`.

I think merge this PR as is, then option 2 could be done later. That's my idea of the OSD too; e.g., when you change the memory size or the video adapter type, the UI will inform you that setting will only be applied after a restart. But it won't restart the emulator every single time you click (or accidentally click) on such a setting... that would be terrible. The situation is a bit similar with option 1 from the CLI, but less pronounced.

Yeah, so I really dislike option 1.

File: src/misc/programs.cpp (hunk 2 above)
Comment (weirddan455) [reply src/misc/programs.cpp]: Sure, changed to use `auto`

**Comment by weirddan455** [comment]
Sounds like a plan @johnnovak 

I'll just merge this as is and we can add that feature in later.

File: src/misc/setup.cpp (hunk 1 above)
Comment (johnnovak) [reply src/misc/setup.cpp:877]: I tend to agree; we don't get brownie points for using STL to the max. Let's optimise for readability.

**Comment by johnnovak** [comment]
I think this is safe to backport too @weirddan455.

**Pull request merged**